enabling doctest-modules causes fixture to not be found

With pytest dev (starting at: 46478fad53774fad76d829f51679824d0131a0b3) the following example passes with a default pytest run but errors if `--doctest-modules` is enabled.

```python
import pytest

@pytest.fixture
def foo_fixture():
    return 1

def test_foo(foo_fixture):
    assert foo_fixture
```

The error is:
```
____________________________________________________________________________________________________________ ERROR at setup of test_foo _____________________________________________________________________________________________________________
file /projects/251117_pytest_dev/foo/test_foo.py, line 10
  def test_foo(foo_fixture):
E       fixture 'foo_fixture' not found
>       available fixtures: _syrupy_apply_ide_patches, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, capteesys, doctest_namespace, monkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, snapshot, subtests, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.

/projects/251117_pytest_dev/foo/test_foo.py:10
```

The prior commit (543b11f2534bb0edc85b86767ed239da9c5bd140) passes in with `--doctest-modules`.

<details>
<summary>pip list</summary>

```
Package           Version                 Editable project location
----------------- ----------------------- ------------------------------------------------
aiohappyeyeballs  2.6.2
aiohttp           3.13.5
aiosignal         1.4.0
asdf              5.3.1.dev86+g71187c0fc  /projects/251117_pytest_dev/asdf
asdf_standard     1.5.0
attrs             26.1.0
frozenlist        1.8.0
fsspec            2026.4.0
idna              3.17
iniconfig         2.3.0
jmespath          1.1.0
lz4               4.4.5
multidict         6.7.1
numpy             2.4.6
packaging         25.0
pip               24.0
pluggy            1.6.0
propcache         0.5.2
psutil            7.2.2
Pygments          2.19.2
pytest            9.1.0.dev386+g46478fad5 /projects/251117_pytest_dev/pytest
PyYAML            6.0.3
semantic-version  2.10.0
syrupy            5.2.0
typing_extensions 4.15.0
yarl              1.24.2
```
</details>

Tested with pytest dev (commits above) mac os.

- [x] a detailed description of the bug or problem you are having
- [x] output of `pip list` from the virtual environment you are using
- [x] pytest and operating system versions
- [x] minimal example if possible

--- a/AUTHORS
+++ b/AUTHORS
@@ -334,6 +334,7 @@ Mike Fiedler (miketheman)
 Mike Hoyle (hoylemd)
 Mike Lundy
 Milan Lesnek
+minbang930
 Miro Hrončok
 Mulat Mekonen
 mrbean-bremen

--- a/changelog/14533.breaking.rst
+++ b/changelog/14533.breaking.rst
@@ -1,0 +1,8 @@
+When using :option:`--doctest-modules`, autouse fixtures with ``module``, ``package`` or ``session`` scope that are defined inline in Python test modules (not plugins or conftests) will now possibly execute twice.
+
+If this is undesirable, move the fixture definition to a ``conftest.py`` file if possible.
+
+Technical explanation for those interested:
+When using `--doctest-modules`, pytest possibly collects Python modules twice, once as :class:`pytest.Module` and once as a ``DoctestModule`` (depending on the configuration).
+Due to improvements in pytest's fixture implementation, if e.g. the ``DoctestModule`` collects a fixture, it is now visible to it only, and not to the ``Module``.
+This means that both need to register the fixtures independently.

--- a/doc/en/how-to/doctest.rst
+++ b/doc/en/how-to/doctest.rst
@@ -42,7 +42,7 @@ By default, pytest will collect ``test*.txt`` files looking for doctest directiv
 can pass additional globs using the :option:`--doctest-glob` option (multi-allowed).
 
 In addition to text files, you can also execute doctests directly from docstrings of your classes
-and functions, including from test modules:
+and functions, including from test modules, using the :option:`--doctest-modules` option:
 
 .. code-block:: python
 
@@ -223,6 +223,9 @@ unless explicitly configured by :confval:`python_files`.
 
 Also, the :ref:`usefixtures <usefixtures>` mark and fixtures marked as :ref:`autouse <autouse>` are supported
 when executing text doctest files.
+
+Python doctest modules are collected independently from Python test files.
+Fixture scope is not shared between the two.
 
 Doctests do not support fixtures that depend on parametrization, because doctest
 collection does not perform the same test generation as normal test functions.

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -377,6 +377,7 @@ testpaths = [
 ]
 norecursedirs = [
     "testing/example_scripts",
+    "testing/plugins_integration",
     ".*",
     "build",
     "dist",

--- a/src/_pytest/doctest.py
+++ b/src/_pytest/doctest.py
@@ -552,8 +552,7 @@ class DoctestModule(Module):
             else:
                 raise
 
-        # While doctests currently don't support fixtures directly, we still
-        # need to pick up autouse fixtures.
+        # doctests supports fixtures via `getfixture` and autouse.
         self.session._fixturemanager.parsefactories(self)
 
         # Uses internal doctest module parsing mechanism.

--- a/src/_pytest/fixtures.py
+++ b/src/_pytest/fixtures.py
@@ -1707,7 +1707,6 @@ class FixtureManager:
         # TODO: The order of the FixtureDefs list of each arg is significant,
         #       explain.
         self._arg2fixturedefs: Final[dict[str, list[FixtureDef[Any]]]] = {}
-        self._holderobjseen: Final[set[object]] = set()
         # A mapping from a node to a list of autouse fixture names it defines.
         # The Session entry holds global usefixtures from config.
         self._node_autousenames: Final[dict[nodes.Node, list[str]]] = {
@@ -2070,8 +2069,6 @@ class FixtureManager:
             assert isinstance(node_or_obj, nodes.Node)
             holderobj = cast(object, node_or_obj.obj)  # type: ignore[attr-defined]
             effective_node = node_or_obj
-        if holderobj in self._holderobjseen:
-            return
 
         # Avoid accessing `@property` (and other descriptors) when iterating fixtures.
         if not safe_isclass(holderobj) and not isinstance(holderobj, types.ModuleType):
@@ -2079,7 +2076,6 @@ class FixtureManager:
         else:
             holderobj_tp = holderobj
 
-        self._holderobjseen.add(holderobj)
         for name in dir(holderobj):
             # The attribute can be an arbitrary descriptor, so the attribute
             # access below can raise. safe_getattr() ignores such exceptions.

--- a/testing/test_doctest.py
+++ b/testing/test_doctest.py
@@ -600,6 +600,37 @@ class TestDoctests:
         reprec = pytester.inline_run(p, "--doctest-modules")
         reprec.assertoutcome(passed=1)
 
+    def test_module_fixture_available_to_normal_test_with_doctestmodules(
+        self, pytester: Pytester
+    ) -> None:
+        """Regression test for #14533.
+
+        Module-level fixtures collected with ``--doctest-modules`` are available
+        both to normal tests and doctests in the same file.
+        """
+        pytester.makepyfile(
+            """
+            import pytest
+
+            @pytest.fixture
+            def fix():
+                return "fix"
+
+            def test(fix):
+                assert fix == "fix"
+
+            def func():
+                '''My function.
+
+                >>> getfixture("fix")
+                'fix'
+                '''
+            """
+        )
+
+        result = pytester.runpytest("--doctest-modules")
+        result.assert_outcomes(passed=2)
+
     def test_doctestmodule_three_tests(self, pytester: Pytester):
         p = pytester.makepyfile(
             """
@@ -1301,6 +1332,37 @@ class TestDoctestAutoUseFixtures:
         )
         result = pytester.runpytest("--doctest-modules")
         result.stdout.fnmatch_lines(["*2 passed*"])
+
+    def test_doctest_and_python_fixtures_not_shared(self, pytester: Pytester) -> None:
+        """Fixture scopes are not shared between doctest and python modules.
+
+        This test is not meant as a hard behavioral test -- sharing scope is
+        also an acceptable behavior (see #14533). But this test ensures and
+        behavior change is done knowingly.
+        """
+        pytester.makepyfile(
+            r"""
+            import pytest
+
+            @pytest.fixture(scope="session", autouse=True)
+            def auto():
+                with open("out", "a", encoding="utf-8") as f:
+                    f.write("RUN\n")
+
+            def test():
+                pass
+
+            def func():
+                '''My function.
+
+                >>> 1 + 1
+                2
+                '''
+        """
+        )
+        result = pytester.runpytest("--doctest-modules")
+        result.assert_outcomes(passed=2)
+        assert Path("out").read_text("utf-8").split() == ["RUN"] * 2
 
     @pytest.mark.parametrize("scope", SCOPES)
     @pytest.mark.parametrize("enable_doctest", [True, False])

--- a/tox.ini
+++ b/tox.ini
@@ -182,6 +182,7 @@ description =
 pip_pre=true
 changedir = testing/plugins_integration
 deps = -rtesting/plugins_integration/requirements.txt
+allowlist_externals = pip
 setenv =
     PYTHONPATH=.
 commands =

